### ISSUE TITLE
Serialization of Yggdrasil models

### ISSUE THREAD

**Issue by JoseAF** [google/yggdrasil-decision-forests#31]
Hi guys

I have come across a couple of methods in the library to LoadModel and SaveModel but I haven't seen any interface provided for serialization of a model to string. I can save to a directory and serialize from there but is there a simpler way to do this? 

Many thanks.

**Comment by JoseAF** [comment]
I've just seen AbstractModel::ExportProto(*mdl, &header). Perhaps this might do the trick. I'll investigate

**Comment by JoseAF** [comment]
I've been looking into the ExportProto/ImportProto interface and have used a couple of the options, to String and to Array, but when parsing back into a model none of the options seem to give me back the full original trained model. I'm doing the following:

TO SERIALIZE TRAINED MODEL INTO A STR
auto* predictionModel = dynamic_cast<gradient_boosted_trees::GradientBoostedTreesModel*>(abstractModel.get());
proto::AbstractModel modelProto;
AbstractModel::ExportProto(*predictionModel, &modelProto);
modelProto.SerializeAsString() => modelStr

TO DESERIALIZE FROM A STR BACK INTO A MODEL
proto::AbstractModel modelProto;
modelProto.ParseFromString(modelStr);
gradient_boosted_trees::GradientBoostedTreesModel predictionModel;
AbstractModel::ImportProto(modelProto, &predictionModel);

Should this give me back the full trained model or does serialization only apply to some of the model metadata but not to the whole model (in which case it's not possible to use the deserialized model for prediction)?

Thanks

**Comment by rstz** [comment]
Hi Jose,

a YDF model is a collection of files, most of them in protobuf format. We do not provide an built-in way to export a model to a string. 

Our Javascript port (see [yggdrasil_decision_forests/port/javascript](https://github.com/google/yggdrasil-decision-forests/tree/main/yggdrasil_decision_forests/port/javascript) ) works with zip-archives containing the model files and this has worked well for us. 

AbstractModel::ExportProto() is not part of the public API of YDF and probably shouldn't be used in production - we may change this function at any time. It also does not do what you plan to do - it only exports a small part of the model (that is described in [yggdrasil_decision_forests/model/abstract_model.proto](https://github.com/google/yggdrasil-decision-forests/tree/main/yggdrasil_decision_forests/model/abstract_model.proto). But a model consists of more files such as the model-specific header and the actual node contents.

**Comment by JoseAF** [comment]
Hi Richard

Thanks a lot for the reply and the clarification regarding the exportproto/importproto interface. It's a pity that there is no public interface for exporting/importing to/from string. 

If I understand correctly, by mentioning the Javascript port you are talking now about saving to files. Should the SaveModel/LoadModel interface not provide this kind of functionality (I have a related thread opened asking this question)? This seems to be a public interface accessed through the model_library lib and in one of the tests provided in Github I see that it's expected that a saved model will have the same full description as a loaded one:

TEST(GradientBoostedTrees, SaveAndLoadModelWithoutPrefix) {
  std::unique_ptr<model::AbstractModel> original_model;
  // TODO: Simplify this test by having it use a toy model.
  EXPECT_OK(model::LoadModel(
      file::JoinPath(TestDataDir(), "model", "adult_binary_class_gbdt"),
      &original_model));
  std::string model_path =
      file::JoinPath(test::TmpDirectory(), "saved_model_without_prefix");
  EXPECT_OK(SaveModel(model_path, original_model.get(), {}));

  std::unique_ptr<model::AbstractModel> loaded_model;
  EXPECT_OK(LoadModel(model_path, &loaded_model, {}));
  EXPECT_EQ(original_model->DescriptionAndStatistics(/*full_definition=*/true),
            loaded_model->DescriptionAndStatistics(/*full_definition=*/true));
}

Am I right in expecting the two models, before and after saving/loading should be exactly the same, and both able to predict and give the same results? I ask because my tests do not show this. Before saving, the trained model has 10 trees and all the data seems to be present. After saving and loading again (this saves for me into 2 .pb files, header.pb and data_spec.pb), the model has some information about variable importance etc, but is missing the actual trees so no predictions are possible. This means, I guess, that the above test would fail?

**Comment by rstz** [comment]
Hi Jose,

_Tl;dr: YDF could provide a way to serialize to a string, but we don't see a use case that would justify the added complexity._

A YDF model is a **folder** containing a variable number of files. Which files a model exactly contains depends on different factors, such as the model type and the model size. Some of the files are in protobuf format, but others are not.

As an example, consider this Random Forest model: [prefixed_adult_binary_class_rf](https://github.com/google/yggdrasil-decision-forests/tree/main/yggdrasil_decision_forests/test_data/model/prefixed_adult_binary_class_rf) You can see that the model contains a [prefixed_header.pb](https://github.com/google/yggdrasil-decision-forests/blob/main/yggdrasil_decision_forests/test_data/model/prefixed_adult_binary_class_rf/prefixed_header.pb), an empty 
 [prefixed_done](https://github.com/google/yggdrasil-decision-forests/blob/main/yggdrasil_decision_forests/test_data/model/prefixed_adult_binary_class_rf/prefixed_done) file, a [prefixed_random_forest_header.pb](https://github.com/google/yggdrasil-decision-forests/blob/main/yggdrasil_decision_forests/test_data/model/prefixed_adult_binary_class_rf/prefixed_random_forest_header.pb) and other files. The only file in this folder that is *not* part of the model and has been added on Github for documentation purposes is [prefixed_description.txt](https://github.com/google/yggdrasil-decision-forests/blob/main/yggdrasil_decision_forests/test_data/model/prefixed_adult_binary_class_rf/prefixed_description.txt). 

Calling `SaveModel(absl::string_view directory, const AbstractModel* const mdl, ModelIOOptions io_options = {})` from [model_library.h](https://github.com/google/yggdrasil-decision-forests/blob/main/yggdrasil_decision_forests/model/model_library.h) with a pointer to a model will create the given directory and all these files. 

The code in your third comment on this issue, however, will only create a subset of these files (namely the header.pb). So it is expected that this code will not work.

The test you pointed out checks that a model that is loaded with LoadModel() and then saved with SaveModel() can be loaded again and will be unchanged. This is always the case (and if it is not, this is a bug and I'd be happy to get a repro). What will not work, however, is take a model, save a subset of files (e.g. using the code you provided in [your third comment](https://github.com/google/yggdrasil-decision-forests/issues/31#issuecomment-1386970657)) and then loading it.

On the larger question of whether or not YDF should provide a built-in way to serialize a model to a string: Adding this functionality is possible, but we do not see the use case for it. In all the use cases that we are aware of, the user is able to just zip (or tar, or base64 encode, or ...) the model directory with a third_party tool / library. Hence, if users need a string, they can create one outside of YDF. Implementing a built-in way in YDF would be non-trivial and therefore add complexity and maintenance load.

**Comment by JoseAF** [comment]
Hi Richard

Thanks for the comments above and sorry to keep going on about this. 

I'm happy to leave to one side the issue of serializing to string. My last question referred to the SaveModel/LoadModel interface which should work by saving the model to a set of files. This is the interface that seems not to be working for me. To describe briefly what I've done:

1. I create an abstract learner configured to be GBT (gradient_boosted_trees_config)
2. I create dataset and dataspec and train the model (TrainWithStatus)
3. Print out DescriptionAndStatistics => all seems good at this point
4. SaveModel (path, pointer to trained abstract model) - this interface is from version 0.2.3 of the YDT lib before prefix functionality was added => this saves only 2 .pb files, header and data_spec, to a folder
5. LoadModel (path, pointer to new abstract model unique pointer)
6. Print out DescriptionAndStatistics => these are missing the actual values of the model trees, weights, etc

From looking at the code, I suspect the problem is with the saving, as I would expect to see at least a third _done_ file in the folder if nothing else. In terms of the model_library, the only change I can see between versions 0.2.3 and 1.2.0 is the addition of the prefix functionality but I don't imagine this has anything to do with the problem I see.

Let me know your thoughts.

**Comment by achoum** [comment]
Thanks for the details and patience.

It is likely that the model is being saved using the C++ `model.Save` method. This is not the correct way to save a model. Please see the documentation of this method for more details. The model should instead be saved using the `SaveModel(model, path)`  function. Note that loading a model (using "LoadModel") that was saved incorrectly (with `model.Save`) will raise an error.

[Here](https://github.com/google/yggdrasil-decision-forests/blob/main/examples/beginner.cc#L127) is an example.

I hope this helps.
